File: scripts/bootstrap.py
import http.client
import netifaces as net
import ipaddress as ips
import http
import os
import base64
import json

adapterPort = os.environ['EMQX_ADAPTER_PORT']
gatePort = os.environ['PORT']
network = os.environ['NETWORK']
emqxHost = os.environ['EMQX_HOST']
emqxPort = os.environ['EMQX_PORT']
emqxUser = os.environ['EMQX_USER']
emqxPass = os.environ['EMQX_PASS']

n = ips.ip_network(network)

for iface in net.interfaces():
    if net.AF_INET in net.ifaddresses(iface):
        for link in net.ifaddresses(iface)[net.AF_INET]:
            addr = ips.ip_address(link['addr'])

            if addr in n.hosts():
                req = {
                    'name': 'exproto',
                    'server': {
                        'bind': adapterPort
                    },
                    'handler': {
                        'address': f'http://{addr}:{gatePort}'
                    }
                }

                http.client.HTTPConnection(f'http://{emqxHost}:{emqxPort}').request(
                    'PUT', '/gateways/exproto', json.dumps(req), {
                        'Content-Type': 'application/json', 
                        'Authorization': f'Basic {base64.b64encode(f'{emqxUser}:{emqxPass}')}'
                    }
                )

                exit(0)

exit(-1)

File: scripts/bootstrap.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"os"
	"os/exec"
)

type Request struct {
	Name   string `json:"name"`
	Server struct {
		Bind string `json:"bind"`
	} `json:"server"`
	Handler struct {
		Addr string `json:"address"`
	} `json:"handler"`
}

func getEnv(env string) string {
	value, ok := os.LookupEnv(env)

	if !ok {
		panic(fmt.Errorf("env not set: %v", env))
	}

	return value
}

func register(ip net.IP) error {
	aPort := getEnv("EMQX_ADAPTER_PORT")
	eHost := getEnv("EMQX_HOST")
	ePort := getEnv("EMQX_PORT")
	eUser := getEnv("EMQX_USER")
	ePass := getEnv("EMQX_PASS")
	gPort := getEnv("PORT")

	cli := &http.Client{}
	url := fmt.Sprintf("http://%s:%s/api/v5/gateways/exproto", eHost, ePort)
	pay, err := json.Marshal(&Request{
		Name: "exproto",
		Server: struct {
			Bind string "json:\"bind\""
		}{
			Bind: aPort,
		},
		Handler: struct {
			Addr string "json:\"address\""
		}{
			Addr: fmt.Sprintf("http://%s:%s", ip.String(), gPort),
		},
	})

	if err != nil {
		return err
	}

	req, err := http.NewRequest(http.MethodPut, url, bytes.NewReader(pay))

	if err != nil {
		return err
	}

	req.SetBasicAuth(eUser, ePass)
	req.Header.Add("Content-Type", "application/json")

	res, err := cli.Do(req)

	if err != nil {
		return err
	}

	defer res.Body.Close()

	if res.StatusCode != 204 {
		msg, err := io.ReadAll(res.Body)

		if err != nil {
			return err
		}

		return errors.New(string(msg))
	}

	return nil
}

func main() {
	_, network, err := net.ParseCIDR(getEnv("NETWORK"))

	if err != nil {
		panic(err)
	}

	addrs, err := net.InterfaceAddrs()

	if err != nil {
		panic(err)
	}

	for _, addr := range addrs {
		if ip, ok := addr.(*net.IPNet); ok && network.Contains(ip.IP) {
			if err := register(ip.IP); err != nil {
				panic(err)
			}

			cmd := exec.Command("haproxy", "-f", "/usr/local/etc/haproxy/haproxy.cfg")

			if err := cmd.Run(); err != nil {
				panic(err)
			}

			return
		}
	}

	panic("host does not belongs to specified network")
}
